README.md
# balloon desktop client

This is the desktop client for the [balloon](https://github.com/gyselroth/balloon) cloud server.

[![Build Status](https://travis-ci.org/gyselroth/balloon-client-desktop.svg?branch=master)](https://travis-ci.org/gyselroth/balloon-client-desktop)
[![GitHub release](https://img.shields.io/github/release/gyselroth/balloon-client-desktop.svg)](https://github.com/gyselroth/balloon-client-desktop/releases)
[![GitHub license](https://img.shields.io/badge/license-GPL-blue.svg)](https://raw.githubusercontent.com/gyselroth/balloon-client-desktop/master/LICENSE) 
 [ ![Download](https://api.bintray.com/packages/gyselroth/balloon/balloon-client-desktop/images/download.svg) ](https://bintray.com/gyselroth/balloon/balloon-client-desktop/_latestVersion) 

Keep your desktop in sync with your data in the cloud.

## Features

* Synchronize your cloud with your desktop!
* Offline support
* Multilingual (and locale based)
* HTTP basic authentication and/or support for (multiple) OpenID-Connect provider
* Integrated auto-update mechanism
* Superfast and operating in the background
* Settings/Status available via tray
* Support for unlink/link instances from multiple balloon server/accounts
* Selective synchronization (Choose what you want)
* Activity log
* NodeJS application, which is compatible with all popular operating systems (e.g. OS X, Windows and Linux)

# Documentation
How to install? Please visit the [documentation](https://gyselroth.github.io/balloon-docs/).

## Requirements
You need a running balloon server. Please see [balloon](https://github.com/gyselroth/balloon) for further information.

## Changelog
A changelog is available [here](https://github.com/gyselroth/balloon-client-desktop/blob/master/CHANGELOG.md).

## Contribute
We are glad that you would like to contribute to this project. Please observe the following [terms](https://github.com/gyselroth/balloon-client-desktop/blob/master/CONTRIBUTING.md).


CHANGELOG.md
## 1.3.0-beta1
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Fri Mar 06 15:39:43 CET 2020

### Bugfixes
* Logger should be able to serialize BigInt #229
* Do not use format.json() in logger #228
* Send feedback to new api endpoint #217
* Trailing slash in server url leads to error #235
* Server can be changed without restarting the client #237
* Unhandled promise rejection if build without releaser #247

### Changes
* Upgrade to electron v7 #132
* Use requestSingleInstanceLock instead of makeSingleInstance #141
* Remove @gyselroth/windows-fsstat dependency #132
* Show selective sync on separate page in startup wizard #172
* Remove windows registry entries on uninstall #32
* Do not unlink instance on network or server errors #168, #188
* Unlink active instance when a user logs in with a different username #231
* If refreshAccessToken fails, don't automaticaly issue auth request #231
* Always load correct instance when login is vreified #232, #231
* Only set client config after correct instance has been loaded #231
* Do not unbind closed event before successful login #211
* Add hint when account has been unlinked by client #171
* Display a meaningful error message if login fails with EPERM #161
* New tray behaviour for linux #207
* Improve sync error messages displayed to users #216
* Use Buffer.from() instead of deprecated new Buffer() #238
* Replace call to deprecated function setTemplateImage #236
* Cancel upload if file has been changed or deleted #140
* Identify instance with user id instead of username #230
* Change wizard finish page message #246

### Packaging
* Add AppImage package and enable autoupdate for linux (If installed using appimage) #248


## 1.2.1
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Thu Jan 09 10:44:43 CET 2020

### Bugfixes
* Use same font-size in textarea as in input field #211
* Improve selective window open time #179
* Display message while update is downloading #59
* replace refreshToken if server issues a new refreshToken #197
* Handle code 20 when moving a share #165

### Deployment
* Windows build fails Error: 403 status code downloading arm64 iojs.lib #223


## 1.2.0
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Fri Nov 29 09:56:11 CET 2019

### Bugfixes
* Missleading message during check for update while downloading latest update #74
* Change of allowPrerelease flag should work without app restart #169
* Soft fail if no rights to send events to system events #193
* Adding missing localizations #205
* Wrong status after link account #205
* Empty Message Prevents Sending Error Report #204
* Fix tray window positioning on osx #112
* Do not unlink account if access token can not get refreshed #188
* Open balloon the 2nd time does not open the balloon folder #202
* Fix initial login state may differ #163

### Sync
* Various sync fixes, balloon-node-sync version v0.5.2

### Deployment
* Submit balloon-desktop to brew #195
* Use travis for windows builds instead appveyor #198

### Common
* Starting with v1.2.0 the changelog has a new format, all changes get splitted into sub categeories within the changelog to get a better readability.


## 1.1.1
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Tue Jul 23 13:14:33 CEST 2019

* CORE [FIX] oidc access token can't be refreshed #200


## 1.1.0
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Wed Jun 12 10:32:21 CEST 2019

no changes


## 1.1.0-beta2
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Tue May 28 15:49:23 CEST 2019

* CORE: [FIX] Failed to connect with server #189
* UI: [FEATURE] Click on a file in tray activities should open parent folder in os file browser #182


## 1.1.0-beta1
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Wed May 22 15:17:23 CEST 2019

* CORE: [FEATURE] Implement token flow and MFA #186
* CORE: [CHANGE] Add os/version to useragent #187


## 1.0.0
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Thu February 07 09:36:23 CET 2019

* CORE: [FIX] Increased tray performance during initial sync #178
* CORE: [CHANGE] Display new tray state while directory structure is synchronized
* CORE: [FIX] Missing client version in user-agent #183
* UI: [FIX] Disable horinzontal scrollbar in tray error page #181


## 1.0.0-rc2
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Thu February 07 09:36:23 CET 2019

* CORE: [FIX] Open selective halts sync #174
* CORE: [FIX] Avoid concurrent refreshAccessToken requests #176
* CORE: [FIX] Avoid creating login item for app translocation paths #177
* CORE: [FIX] Corrupt Feedback #175
* CORE: [FIX] Unbind lsiteners before binding them again #178


## 1.0.0-rc1
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Fri December 21 16:27:35 CET 2018

* UI: [FIX] Small user interface fixes
* UI: [FIX] balloon path cut off #170
* CORE: [FIX] Restart full sync after sync received 401 and refreshed acess token


## 1.0.0-beta4
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Fri December 21 10:29:35 CET 2018

* UI: [FIX] Small user interface fixes
* UI: [FIX] Tray sync continue not updated back to sync pause after click #166
* UI: [FIX] balloon folder not changed in ui after selecting a new one during startup wizard #167
* SYNC: [CHANGE] Upgrade sync to v0.3.0-beta2


## 1.0.0-beta3
**Maintainer**: balloon-team <opensource@gyselroth.net>\
**Date**: Thu December 20 14:29:34 CET 2018

* CORE: [FIX] Upgrade from 0.2.4 to 1.0.0-beta1 unlinked active instance #160
* CORE: [FIX] TypeError: this.app.whenReady is exception #158
* CORE: [FIX] Instances not in feedbacks metadata #157
* CORE: [FIX] Ignore shares by default, open selective before sync added new share to ignore db #162
* UI: [FIX] Used Storage Percentage #159
* UI: [FIX] Fixed de-CH and en-US translations
* SYNC: [CHANGE] Upgrade sync to v0.3.0-beta1


## 1.0.0-beta2

app/main.js
const electron = require('electron');
const {app, ipcMain, dialog} = require('electron');

const env = require('./env.js');
const clientConfig = require('./lib/config.js');
const appState = require('./lib/state.js');
const migrate = require('./lib/migrate.js');
const TrayCtrl = require('./ui/tray/controller.js');
const SelectiveCtrl = require('./ui/selective/controller.js');
const SyncCtrl = require('./lib/sync/controller.js');
const StartupCtrl = require('./ui/startup/controller.js');
const AuthCtrl = require('./lib/auth/controller.js');
const AutoUpdateCtrl = require('./lib/auto-update/controller.js');
const FeedbackCtrl = require('./ui/feedback/controller.js');
const setMenu = require('./lib/menu.js');
const BalloonDirSelectorCtrl = require('./ui/balloon-dir-selector/controller.js');
const {BalloonBurlHandler} = require('./lib/burl.js');
const i18n = require ('./lib/i18n.js');

const logger = require('./lib/logger.js');
const loggerFactory = require('./lib/logger-factory.js');
const configManager = require('./lib/config-manager/controller.js')(clientConfig);
const globalConfig = require('./lib/global-config.js');
var isHandlingUnauthorizedRequest = false;
var isHandlingMfaRequired = false;

var tray, selective, sync, feedback, autoUpdate;

var standardLogger = new loggerFactory(clientConfig.getAll());
var startup = StartupCtrl(env, clientConfig);
var auth = AuthCtrl(env, clientConfig);
var selective = SelectiveCtrl(env, clientConfig);
var ballonDirSelector = BalloonDirSelectorCtrl(env, clientConfig);
var burlHandler = new BalloonBurlHandler(clientConfig);

if (process.platform === 'linux') {
  app.disableHardwareAcceleration();
}

logger.setLogger(standardLogger);

function extractBurlArgument(argv) {
  return argv.find(argument => {return burlHandler.isBalloonBurlPath(argument)});
}

process.on('uncaughtException', function(exception) {
  logger.error('uncaught exception', {
    category: 'main',
    error: exception
  });
});

function openBurl(burlPath) {
  if (burlHandler.isBalloonBurlPath(burlPath)) {
    burlHandler.extractBurl(burlPath).then((burl) => {
      dialog.showMessageBox(null, {
        type: 'question',
        buttons: [i18n.__('button.continue'), i18n.__('button.cancel')],
        title: 'Balloon URL',
        message: i18n.__('burl.prompt'),
        detail: burl,
        cancelId: 1,
      }).then((result) => {
        if (0 === result.response) {
          burlHandler.handleBurl(burl);
        }
      });
    }).catch((error) => {
      dialog.showMessageBox(null, {
        type: 'error',
        buttons: [i18n.__('button.close')],
        title: 'Balloon URL',
        message: i18n.__('burl.' + error.error),
        detail: error.burl,
      });
    });
  } else {
    logger.warn('Tried to open an invalid burl path', {burlPath});
  }
}

function startApp() {
  logger.info('bootstrap app', {category: 'main'});

  auth.retrieveLoginSecret().then(() => {
    logger.info('login secret recieved', {category: 'main'});

    ipcMain.once('tray-online-state-changed', function(event, state) {
      tray.toggleState('offline', !state);
      bindOnlineStateChanged();
      if(clientConfig.hadConfig()) {
        tray.create();
        autoUpdate.checkForUpdate();
      }

      logger.info('initial online state', {
        category: 'main',
        state: state
      });

      appState.set('onLineState', state);
      startup.checkConfig().then((result) => {
        logger.info('startup checkconfig successfull', {
          category: 'main',
        });

        if(!tray.isRunning()) {
          tray.create();
        }

        sync = SyncCtrl(env, tray);

        const welcomeWizard = result[2] === undefined || !result[2].welcomeWizardPromise ? Promise.resolve() : result[2].welcomeWizardPromise;

        welcomeWizard.then(initializeSync);
      }).catch((error) => {
        switch(error.code) {
          case 'E_BLN_CONFIG_CREDENTIALS':
            logger.warning('startup checkconfig no credentials, unlinking account', {category: 'main', error: error});
            unlinkAccount(true);
          break;
          default:
            if(!handleAuthError('start-app', error)) {
              logger.error('startup checkconfig error, quitting app', {category: 'main', error: error});
              app.quit();
            }
          break;
        }
      });
    });

    tray = TrayCtrl(env, clientConfig);
    autoUpdate = AutoUpdateCtrl(env, clientConfig, tray);
  });
}

function initializeSync() {
  if(sync && sync.getMayStart() === false || !sync) {
    electron.powerMonitor.on('suspend', () => {
      logger.info('the system is going to sleep', {
        category: 'main',
      });

      //abort a possibly active sync if not already paused
      if(sync && sync.isPaused() === false) {
        sync.pause(true);
      } else if(sync) {
        //if sync is paused kill a possible active watcher
        sync.killWatcher();
      }
    });

    electron.powerMonitor.on('resume', () => {
      logger.info('the system is resuming', {
        category: 'main',
      });

      startSync(true);
    });
  }

  if(!sync) {
    sync = SyncCtrl(env, tray);
  }

  sync.setMayStart(true);
  if(sync.isPaused() === false) startSync(true);
}

function unlinkAccount(clientInitiated) {
  setDisconnectedState(false);

  (function() {
    if(!sync) return Promise.resolve();

    return sync.pause(true);
  }()).then(function() {
    return auth.logout(clientInitiated);
  }).then(() => {
    logger.info('logout successfull', {
      category: 'main',
    });

    tray.emit('unlink-account-result', true);
    tray.toggleState('loggedout', true);
  }).catch((error) => {
    logger.error('logout not successfull', {
      category: 'main',
      error: error
    });

    tray.emit('unlink-account-result', false);
  });
}

function handleUnauthorizedRequest() {
  return new Promise(function(resolve, reject) {
    logger.debug('got 401, refresh accessToken', {category: 'main'});

    auth.refreshAccessToken().then(resolve).catch(err => {
      if(!handleAuthError('hanlde-unauthorized-request', err)) {
        logger.error('could not refresh accessToken, unlink instance', {category: 'main', err});
        unlinkAccount(true);
      }

      reject();
    });
  });
}

function handleAuthError(subcategory, error) {
  if(!error || !error.code) return false;

  let authErrorHandled = true;
  switch(error.code) {
    case 'E_BLN_AUTH_NETWORK':
    case 'E_BLN_OIDC_NETWORK':
    case 'E_BLN_AUTH_SERVER':
    case 'E_OIDC_AUTH_SERVER':
      logger.info('Auth server responded with a invalid status', {category: 'main', subcategory});
      setDisconnectedState(true);
    break;
    default:
      authErrorHandled = false;
    break;
  }

  return authErrorHandled;
}

function bindOnlineStateChanged() {
  ipcMain.on('tray-online-state-changed', function(event, state) {
    logger.info('online state changed', {
      category: 'main',
      state: state,
      syncPaused: (sync && sync.isPaused()),
      needsStartup: startup.needsStartupWizzard(),
      loggedIn: clientConfig.get('loggedin')
    });

    appState.set('onLineState', state);
    if(state === false) {
      //abort a possibly active sync if not already paused
      if(sync && sync.isPaused() === false) sync.pause(true);
      tray.toggleState('offline', true);
    } else {
      tray.toggleState('offline', false);
      if(clientConfig.get('loggedin') === true) {
        if(sync && sync.isPaused() === false) startSync(true);
      } else if(startup.needsStartupWizzard() === false) {
        tryToReconnect();
      }
    }
  });
}

function tryToReconnect() {
  logger.debug('Trying to verify user credentials', {category: 'main'});

  // Pass rejected promise, as it should silently fail
  auth.login(() => Promise.reject()).then(() => {
    clientConfig.updateTraySecret();
    tray.toggleState('loggedout', false);
    setDisconnectedState(false);
    initializeSync();
  }).catch((error) => {
    if(!handleAuthError('online-state-changed-event', error)) {
      logger.warning('User is not authenticated', {category: 'main', error});
      tray.emit('unlink-account-result', true);
      tray.toggleState('loggedout', true);
    }
  });
}

function setDisconnectedState(state) {
  tray.toggleState('disconnected', state);
  appState.set('disconnected', state);

  if(state === true) {
    tray.emit('disconnected');
  } else {
    tray.emit('connected');
  }
}

var gotLock = app.requestSingleInstanceLock();

if(gotLock === false) {
  app.quit();
} else {
  app.on('open-file', function(event, path) {
    if (process.platform === 'darwin') {
      logger.info('OPEN BURL open-file', {
          category: 'main',
          path
      });
      openBurl(path);
    }
  });

  app.on('second-instance', function(event, argv, workingDirectory) {
    let burlArgument = extractBurlArgument(argv);
    if (burlArgument) {
      logger.info('OPEN BURL sec instance', {
          category: 'main',
          burlArgument
      });
      openBurl(burlArgument);
    } else {
      if (tray.isWindowVisible() || process.platform !== 'linux') {
        startup.showBalloonDir();
      } else {
        tray.show();
      }
    }
  });

  app.on('ready', function () {
    appState.set('updateAvailable', false);
    appState.set('updateDownloading', false);
    appState.set('disconnected', false);

    feedback = FeedbackCtrl(env, clientConfig);
    feedback.toggleAutoReport(globalConfig.get('autoReport'));

    logger.info('app ready to operate', {
        category: 'main',
        execPath: process.execPath,
        appPath: app.getAppPath()
    });

    setMenu();

    migrate().then(result => {
      startApp();
    }).catch(err => {
      logger.error('error during migration, quitting app', {
        category: 'main',
        error: err
      });

      app.quit();
    })
  });
}
/** Main App **/
ipcMain.on('quit', function() {
  app.quit();
});

/** Tray **/
ipcMain.on('tray-toggle', function() {
  tray.toggle();
});

ipcMain.on('tray-hide', function() {
  tray.hide();
});

ipcMain.on('tray-show', function() {
  tray.show();
});

/** **/
ipcMain.on('try-to-reconnect', () => {
  logger.info('trying to reconnect', {category: 'main'});

  tryToReconnect();
});

/** Settings **/
ipcMain.on('settings-autoReport-changed', function(event, state) {
  feedback.toggleAutoReport(state);
});

/** Auto update **/
ipcMain.on('install-update', function() {
  logger.info('install-update triggered', {
    category: 'main',
  });

  dialog.showMessageBox(null, {
    type: 'question',
    buttons: [i18n.__('button.continue'), i18n.__('button.cancel')],
    title: 'Update',
    message: i18n.__('update.install'),
  }, (buttonIndex) => {
    if (0 === buttonIndex) {
      autoUpdate.quitAndInstall();
    }
  });
});

ipcMain.on('check-for-update', function() {
  logger.info('check-for-update triggered', {
      category: 'main',
  });

  autoUpdate.checkForUpdate();
});

/** Sync **/
ipcMain.on('sync-toggle-pause', () => {
  if(!sync) {
    return tray.syncPaused();
  }

  tray.toggleState('pause', !sync.isPaused());
  sync.togglePause();
});

ipcMain.on('selective-open', function(event) {
  selective.open();
});

ipcMain.on('selective-close', function(event) {
  selective.close();
});

ipcMain.on('selective-apply', function(event, difference) {
  logger.info('Applying selective sync changes', {category: 'main', difference});

  if(sync) sync.updateSelectiveSync(difference, err => {
    selective.close();
  });
});

ipcMain.on('selective-ignore-new-shares', function(event) {
  if(sync) {
    sync.ignoreNewShares((err, ) => {
      event.sender.send('selective-ignore-new-shares-result');
    });
  } else {
    event.sender.send('selective-ignore-new-shares-result');
  }
});

ipcMain.on('balloonDirSelector-open', function(event) {
  (function() {
    if(!sync) return Promise.resolve();

    return sync.pause(true);
  }()).then(function() {
    return ballonDirSelector.open();
  }).then((result) => {
    event.sender.send('balloonDirSelector-result', result);
    startSync((result && result.newPath));
  }).catch((err) => {
    if(err) logger.error('Change balloon dir failed', {category: 'main', err});
    startSync(false);
  });
});

ipcMain.on('unlink-account', (event) => {
  logger.info('logout requested', {category: 'main'});
  unlinkAccount(false);
});

ipcMain.on('link-account', (event, id) => {
  logger.info('login requested', {category: 'main'});

  startup.checkConfig().then(() => {
    logger.info('login successfull', {
      category: 'main',
      data: clientConfig.getMulti(['username', 'loggedin'])
    });

    clientConfig.updateTraySecret();

    tray.toggleState('loggedout', false);
    setDisconnectedState(false);
    initializeSync();
    tray.emit('link-account-result', true);
  }).catch((error) => {
    switch(error.code) {
      case 'E_BLN_OAUTH_WINDOW_OPEN':
        logger.info('login aborted as there is already a login window open', {
          category: 'main',
        });
      break;
      default:
        if(!handleAuthError('link-account-event', error)) {
          logger.error('login not successfull', {category: 'main', error});
          tray.emit('link-account-result', false);
        } else {
          dialog.showMessageBox(null, {
            type: 'warning',
            buttons: [i18n.__('button.ok')],
            title: i18n.__('error.auth'),
            message: i18n.__('error.auth.network_server_error'),
          });
        }
      break;
    }
  });
});

ipcMain.on('selective-error', (event, error, url, line, message) => {
  switch(error.code) {
    case 'E_BLN_API_REQUEST_UNAUTHORIZED':
      selective.close();

      handleUnauthorizedRequest().then(() => {
        selective.open();
      }).catch(() => {
        logger.warn('Could not open selective sync, user not authenticated.', {category: 'main'});
      });
    break;
    case 'E_BLN_API_REQUEST_MFA_REQUIRED':
      selective.close();
      logger.info('got 403 MFA required from selective, unlink account', {category: 'main'});
      unlinkAccount(true);
    break;
    default:
      logger.error('Uncaught selective error.', {
        category: 'main',
        error,
        url,
        line,
        errorMsg: message
      });

      selective.close();
  }
});

ipcMain.on('sync-error', (event, error, url, line, message) => {
  switch(error.code) {
    case 'E_BLN_API_REQUEST_UNAUTHORIZED':
      if(isHandlingUnauthorizedRequest === true) return;

      isHandlingUnauthorizedRequest = true;
      endSync();
      if(sync) sync.killWatcher();


      handleUnauthorizedRequest().then(() => {
        startSync(true);
        isHandlingUnauthorizedRequest = false;
      }).catch(() => {
        isHandlingUnauthorizedRequest = false;
      });
    break;
    case 'E_BLN_API_REQUEST_MFA_REQUIRED':
      if(isHandlingMfaRequired === true) return;

      isHandlingMfaRequired = true;

      logger.info('got 403 MFA required from selective, unlink account', {category: 'main'});
      unlinkAccount(true);
      isHandlingMfaRequired = false;
    break;
    case 'E_BLN_CONFIG_CREDENTIALS':
      logger.error('credentials not set', {
        category: 'main',
        code: error.code
      });

      endSync();
      unlinkAccount(true);
    break;
    case 'E_BLN_CONFIG_BALLOONDIR':
    case 'E_BLN_CONFIG_CONFIGDIR':
    case 'E_BLN_CONFIG_CONFIGDIR_NOTEXISTS':
    case 'E_BLN_CONFIG_APIURL':
      //this should only happen, when user deletes the configuation, while the application is running
      logger.info('reinitializing config, config sync error occured', {
        category: 'main',
        code: error.code
      });

      clientConfig.initialize();
      endSync();
      startSync(true);
    break;
    case 'E_BLN_CONFIG_CONFIGDIR_ACCES':
      logger.error('config dir not accesible.', {
        category: 'main',
        error
      });
      endSync();
    break;
    case 'ENOTFOUND':
    case 'ETIMEDOUT':
    case 'ENETUNREACH':
    case 'EHOSTUNREACH':
    case 'ECONNREFUSED':
    case 'EHOSTDOWN':
    case 'ESOCKETTIMEDOUT':
    case 'ECONNRESET':
      logger.error('sync terminated with networkproblems.', {
        category: 'main',
        code: error.code
      });
      endSync();
      tray.emit('network-offline');
    break;
    case 'BLN_API_DELTA_RESET':
      logger.info('Remote delta triggered reset. Resetting cursor and db', {
        category: 'main',
        code: error.code
      });

      endSync();
      configManager.resetCursorAndDb().then(function() {
        startSync(true);
      }).catch(function(err) {
        startSync(true);
      });
    break;
    default:
      logger.error('Uncaught sync error. restarting sync', {
        category: 'main',
        error,
        url,
        line,
        errorMsg: message
      });

      endSync();
      startSync(true);
  }
});

/** Development Methods **/
if(env.name === 'development') {
  process.on('unhandledRejection', r => console.log(r));
}

if (process.platform === 'darwin' && app.dock && env.name === 'production') {
  //hide from dock on OSX in production
  app.dock.hide();
}

function startSync(forceFullSync) {
  logger.debug('start sync requested', {category: 'main', forceFullSync});

  if(!sync) {
    sync = SyncCtrl(env, tray);
  }

  if(appState.get('onLineState') === true) {
    sync.start(forceFullSync);
  } else {
    logger.info('not starting Sync because client is offline', {
      category: 'main',
    });
  }
}

function endSync() {
  if(sync) sync.endFullSync();
}
